Split `Ray` into `Ray2d` and `Ray3d` and simplify plane construction

Jondolf opened the pull request:
# Objective

A better alternative version of #10843.

Currently, Bevy has a single `Ray` struct for 3D. To allow better interoperability with Bevy's primitive shapes (#10572) and some third party crates (that handle e.g. spatial queries), it would be very useful to have separate versions for 2D and 3D respectively.

## Solution

Separate `Ray` into `Ray2d` and `Ray3d`. These new structs also take advantage of the new primitives by using `Direction2d`/`Direction3d` for the direction:

```rust
pub struct Ray2d {
    pub origin: Vec2,
    pub direction: Direction2d,
}

pub struct Ray3d {
    pub origin: Vec3,
    pub direction: Direction3d,
}
```

and by using `Plane2d`/`Plane3d` in `intersect_plane`:

```rust
impl Ray2d {
    // ...
    pub fn intersect_plane(&self, plane_origin: Vec2, plane: Plane2d) -> Option<f32> {
        // ...
    }
}
```

---

## Changelog

### Added

- `Ray2d` and `Ray3d`
- `Ray2d::new` and `Ray3d::new` constructors
- `Plane2d::new` and `Plane3d::new` constructors

### Removed

- Removed `Ray` in favor of `Ray3d`

### Changed

- `direction` is now a `Direction2d`/`Direction3d` instead of a vector, which provides guaranteed normalization
- `intersect_plane` now takes a `Plane2d`/`Plane3d` instead of just a vector for the plane normal
- `Direction2d` and `Direction3d` now derive `Serialize` and `Deserialize` to preserve ray (de)serialization

## Migration Guide

`Ray` has been renamed to `Ray3d`.

### Ray creation

Before:

```rust
Ray {
    origin: Vec3::ZERO,
    direction: Vec3::new(0.5, 0.6, 0.2).normalize(),
}
```

After:

```rust
// Option 1:
Ray3d {
    origin: Vec3::ZERO,
    direction: Direction3d::new(Vec3::new(0.5, 0.6, 0.2)).unwrap(),
}

// Option 2:
Ray3d::new(Vec3::ZERO, Vec3::new(0.5, 0.6, 0.2))
```

### Plane intersections

Before:

```rust
let result = ray.intersect_plane(Vec2::X, Vec2::Y);
```

After:

```rust
let result = ray.intersect_plane(Vec2::X, Plane2d::new(Vec2::Y));
```

alice-i-cecile reviewed (changes requested):
I really like the split: more consistent and efficient.

I think these `From` impls should be cut though.

Jondolf commented on crates/bevy_math/src/primitives/dim2.rs:
```diff
@@ -63,6 +64,14 @@ pub struct Plane2d {
 }
 impl Primitive2d for Plane2d {}
 
+impl From<Vec2> for Plane2d {
```
Should `intersect_plane` take a `Plane2d` then? Previously it just took the plane normal as a vector, but if we have plane primitives, it'd make sense to use them. In that case, we should probably also add a `new` constructor for planes to make the API less painful:

```rust
// No constructor
let result = ray.intersect_plane(
    Vec2::X,
    Plane2d {
        normal: Direction2d::new(Vec2::Y).unwrap(),
    },
);

// Constructor (should probably panic for invalid Direction2d after #10857)
let result = ray.intersect_plane(Vec2::X, Plane2d::new(Vec2::Y));

// Current API
let result = ray.intersect_plane(Vec2::X, Vec2::Y);
```

alice-i-cecile replied:
Agreed on both counts!

Jondolf commented:
I removed the `From` impls from the planes, added `new` constructors instead, and updated the PR description. I think #10857 could be merged first though since the constructors will have to be changed a bit

Jondolf commented:
I updated the PR to match the direction constructor changes. I opted to make the `new` constructors for rays and planes panic with an `expect` if the direction is correct, since I think that's nicer for ergonomics, and you could just not use the constructor if you wanted to handle those cases manually.

If the changes look fine, everything should now be ready :)

rlidwka commented:
How common is it to convert between Ray2 to Ray3? Do we need helper methods for that like glam's `extend()` and `truncate()`?

Jondolf commented:
I don't think it's particularly common, but it'd probably be nice to have. But it would also make sense for other types, like directions, planes, lines, line segment and even rectangle/cuboid, so it should be done in a separate PR.
